feat: add input validation and error handling to GetBookByID

vamshi1188 opened the pull request:
This PR improves the GetBookByID method in the BookModel by adding:

✅ Input validation to ensure bookID is not empty

✅ Proper error handling for SQL errors

✅ Graceful handling of sql.ErrNoRows when no matching book is found

✅ Contextual error messages for easier debugging

These changes make the function safer, more robust, and production-ready.

SaketV8 commented:
Hey @vamshi1188, I have implemented the following sample:

- **For handling the specific case when no book is found**
  I have implemented it in the `GetBookByIDHandler` function itself, so that the query to the DB function looks clean and we can modify our output in the handler itself.

Also,
**It handles the case of an empty ID by**
returning the `404 page not found` if no ID is present in the URL.

And for an empty ID provided like `""`, it will return:

```
{
  "details": "sql: no rows in result set",
  "error": "Failed to get book by ID"
}
```
![image](https://github.com/user-attachments/assets/ca47daba-8ab8-45af-8e76-d6ae04cdbab3)



**Thank you for bringing to my attention that I wasn’t returning the specific ID of the book that couldn’t be found**

khareyash05 commented on book-store-inventory/pkg/repositories/book_repository.go:
```diff
@@ -50,7 +51,12 @@ func (bookModel *BookModel) GetBooks() ([]models.Book, error) {
-	statement := `SELECT book_id, title, author, price, published_date, stock_quantity FROM books WHERE book_id = ? ;`
+	// Input validation
+	if bookID == "" {
+		return models.Book{}, fmt.Errorf("book ID cannot be empty")
```
we should handle custom errors if any param is empty, maybe return a model which say things are empty so that it can be handled on the handler side

vamshi1188 commented:
Thanks for the valuable suggestions, @SaketV8 and @khareyash05.

Here's what's now handled:

    Added custom error handling for empty bookID in the repository layer and used that to return clear messages from the handler.

    Improved the handler logic to return detailed errors (like sql.ErrNoRows) gracefully.

    Adjusted the route to *book_id to properly catch edge cases like /id/ or missing values.

    Added validations to differentiate between:

         Empty path param → returns Invalid or empty book ID

         Non-existent ID → returns Book not found

         Valid ID → returns book details

Tested all 3 scenarios via Postman, and everything works as expected now.
![Screenshot from 2025-06-18 20-27-49](https://github.com/user-attachments/assets/05cc841c-89d0-49af-9b7c-84542df5ab00)
![Screenshot from 2025-06-18 20-27-35](https://github.com/user-attachments/assets/b1f6aadb-4d80-4e70-9eb1-596fd5899e09)
![Screenshot from 2025-06-18 20-27-17](https://github.com/user-attachments/assets/1fc63b9d-2484-4049-ba42-ab9190ac6d64)

vamshi1188 commented:
@khareyash05   i made changes van you please review it 
thank you!

khareyash05 commented:
Apologies for delay in reivew @vamshi1188
